controllers/mysqlcluster_controller.go

```diff
@@ -0,0 +1,476 @@
+	}
+	return bytes, nil
+}
+
+func (r *MySQLClusterReconciler) createOrUpdateHeadlessService(ctx context.Context, log logr.Logger, cluster *mysov1alpha1.MySQLCluster) error {
+	headless := &corev1.Service{}
+	headless.SetNamespace(cluster.Namespace)
+	headless.SetName(cluster.Name)
+
+	op, err := ctrl.CreateOrUpdate(ctx, r.Client, headless, func() error {
+		headless.Labels = map[string]string{
+			appNameKey: fmt.Sprintf("%s-%s", appName, cluster.Name),
+		}
+		headless.Spec.ClusterIP = corev1.ClusterIPNone
+		headless.Spec.Selector = map[string]string{
+			appNameKey: fmt.Sprintf("%s-%s", appName, cluster.Name),
+		}
+		return ctrl.SetControllerReference(cluster, headless, r.Scheme)
+	})
+	if err != nil {
+		log.Error(err, "unable to create-or-update headless Service")
+		return err
+	}
+
+	if op != controllerutil.OperationResultNone {
+		log.Info("reconcile headless Service successfully", "op", op)
+	}
+	return nil
+}
+
+func (r *MySQLClusterReconciler) createOrUpdateRBAC(ctx context.Context, log logr.Logger, cluster *mysov1alpha1.MySQLCluster) error {
+	sa := &corev1.ServiceAccount{}
+	sa.SetNamespace(cluster.Namespace)
+	sa.SetName(cluster.Name)
+
+	op, err := ctrl.CreateOrUpdate(ctx, r.Client, sa, func() error {
+		return nil
+	})
+
+	if err != nil {
+		log.Error(err, "unable to create-or-update ServiceAccount")
+		return err
+	}
+
+	if op != controllerutil.OperationResultNone {
+		log.Info("reconcile ServiceAccount successfully", "op", op)
+	}
+	return nil
+}
+
+func (r *MySQLClusterReconciler) createOrUpdateStatefulSet(ctx context.Context, log logr.Logger, cluster *mysov1alpha1.MySQLCluster) error {
+	sts := &appsv1.StatefulSet{}
+	sts.SetNamespace(cluster.Namespace)
+	sts.SetName(cluster.Name)
+
+	op, err := ctrl.CreateOrUpdate(ctx, r.Client, sts, func() error {
+		sts.Labels = map[string]string{
+			appNameKey: fmt.Sprintf("%s-%s", appName, cluster.Name),
+		}
+		sts.Spec.Replicas = &cluster.Spec.Replicas
+		sts.Spec.PodManagementPolicy = appsv1.ParallelPodManagement
+		sts.Spec.ServiceName = cluster.Name
+		sts.Spec.Selector = &metav1.LabelSelector{}
+		sts.Spec.Selector.MatchLabels = map[string]string{
+			appNameKey: fmt.Sprintf("%s-%s", appName, cluster.Name),
+		}
+		sts.Spec.Template = r.makePodTemplate(log, cluster)
+		sts.Spec.VolumeClaimTemplates = append(
+			r.makeVolumeClaimTemplates(cluster),
+			r.makeDataVolumeClaimTemplate(cluster),
+		)
+		return ctrl.SetControllerReference(cluster, sts, r.Scheme)
+	})
+	if err != nil {
+		log.Error(err, "unable to create-or-update StatefulSet")
+		return err
+	}
+
+	if op != controllerutil.OperationResultNone {
+		log.Info("reconcile StatefulSet successfully", "op", op)
+	}
+	return nil
+}
+
+func (r *MySQLClusterReconciler) makePodTemplate(log logr.Logger, cluster *mysov1alpha1.MySQLCluster) corev1.PodTemplateSpec {
+	template := cluster.Spec.PodTemplate
+	newTemplate := corev1.PodTemplateSpec{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels:      template.Labels,
```

Comment:
You need to shallow copy `template.Labels` because it is later modified.

Reply:
Please check if there are similar problems.

Reply:
> You need to shallow copy template.Labels because it is later modified.

This means that we need to use deep copy?

Reply:
@binoue 
No, shallow copy is correct.
Deep copy is recursive, while shallow copy is not.

I suggested a fix already.